app/forms/reset_password_form.rb

```diff
@@ -30,7 +30,7 @@ def valid_token
       # If the user is not saved in the database, that means looking them up by
       # their token failed
       errors.add(:reset_password_token, 'invalid_token', type: :invalid_token)
-    elsif !user.reset_password_period_valid?
+    elsif !user.reset_password_period_valid? || invalid_account?
```

Comment:
Also worth noting that this should resolve the vast vast majority of cases, but the chance still exists that the two accounts could end up in a race condition where they both try to confirm the email address at the same time in the window of time between the `SELECT` here and the `UPDATE` that follows.